Allow users to override kernels for existing C++ ops through Python

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76250
* __->__ #75905

API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.Library("aten", "IMPL")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
assert torch.sum(x) == x
del my_lib1
assert torch.sum(x) == torch.tensor(3)
```

anjali411 committed:
Allow users to override kernels for existing C++ ops through Python

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 commented on test/test_python_dispatch.py:
```diff
@@ -10,6 +10,52 @@
+
+        # Now we are secretly making the operator a view op so autograd needs to know how
+        # to handle it
+        torch.ops.aten.neg.default.impl("AutogradCPU", my_neg)
```
How do I know I should use "AutogradCPU" here?
Alban: you need more info about dispatcher.

anjali411 replied:
Ed: Is there a way to distinguish if I am adding a new kernel for a dispatch key which didn't have any impl or overriding an existing kernel (wanna use it for adding support for meta tensors)

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

ezyang commented on torch/__init__.py:
```diff
@@ -47,7 +48,7 @@
     'is_deterministic_algorithms_warn_only_enabled',
     'set_deterministic_debug_mode', 'get_deterministic_debug_mode',
-    'set_warn_always', 'is_warn_always_enabled',
+    'set_warn_always', 'is_warn_always_enabled', 'extend_library'
```
A thought here: should we be putting this function in the top-level torch namespace? An alternative is to place it in `torch.ops` (hazard: `extend_library` would then no longer be a valid namespace--though it's hard to imagine anyone wanting to write a namespace this way), `torch.overrides` (this is not quite right, as overrides is about `torch_function` level override and that's not what you're doing here) or `torch.library` (this is symmetric with the C++ header `torch/library.h`).

I think we should seriously consider putting this in `torch.library`

anjali411 replied:
> torch.library (this is symmetric with the C++ header torch/library.h)

That's a very reasonable suggestion and makes a lot of sense. In general, what are some general rules to assess if something should be added to the `torch` namespace? For example, how do we decide if we should only expose it in `torch.library` or expose it in both `torch` and `torch.library`?

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+import torch
+
+existing_libraries = {'aten', 'quantized'}
```
This doesn't seem like the right strategy for matching against existing libraries. In particular, if you are loading a C++ extension, you could end up with extra libraries that are preexisting. Instead, it seems like we should rely on the C++ itself on preventing duplicate libraries (and I think we already have the logic to do this)

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -50,6 +54,70 @@ inline torch::CppFunction dispatch_str(const char* key, Func&& raw_f) {
+  // libtorch_python, so we want to disarm the deallocation if that happens.
+  // PyInterpreter does this correctly, pybind11 does not.
+  ~PythonKernelHolder() {
+    getPyInterpreter()->decref(func_->ptr(getPyInterpreter()), /*is_tensor*/ false);
```
You shouldn't need this anymore; SafePyObject should be stored directly by value and its destructor will take care of this logic for you. (This comment is mooted by the one above, but I feel also obligated to mention: this current code leaks SafePyObject as every new has to be paired with a delete.)

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -50,6 +54,70 @@ inline torch::CppFunction dispatch_str(const char* key, Func&& raw_f) {
+
+    // TODO: actually populate kwargs sometimes?  At the moment, every argument
+    // just gets passed positionally
+    py::dict kwargs;
```
So, it would be very good to deduplicate this code with torch/csrc/autograd/python_variable.cpp. In particular, we actually fixed the logic in python_variable.cpp to properly put kwarg only arguments in the kwargs dict. Since this is a copy paste you didn't get that fix. Putting this all in a helper function would solve that problem.

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -50,6 +54,70 @@ inline torch::CppFunction dispatch_str(const char* key, Func&& raw_f) {
+      auto outs = py::cast<py::sequence>(out);
+      for (unsigned idx = 0; idx < outs.size(); idx++) {
+        torch::jit::push(stack, torch::jit::toIValue(outs[idx].ptr(), op.schema().returns()[idx].type()));
+      }
```
I think there were also some fixes here which haven't made it to this copy of the code. (Helper function plz)

ezyang commented on torch/_ops.py:
```diff
@@ -65,6 +65,15 @@ def overloadpacket(self):
+    def impl(self, lib_obj, dispatch_key, fn):
+        name = self.__name__ if self._overloadname != 'default' else self.__name__.split(".")[0]
+        lib_obj.impl(name, dispatch_key, fn)
+        return
```
We were going to get rid of this helper right?

ezyang replied:
nit: trailing return is unnecessary

anjali411 replied:
yeah we are

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+    # kind can be DEF, FRAGMENT (in C++ IMPL)
+    def __init__(self, kind, ns, dispatch_key=""):
+       # also prevent users from creating libraries with existing C++ libraries, e.g., aten
+       if ns in existing_libraries and kind == "DEF":
```
please fix the lint lol, it makes the diff kind of unreadable haha

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+import torch
```
Put an `__all__` on this module to control visibility

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+# libraries calling into kernels not intended to be called.
+impls = set()
+
+class _Library:
```
not sure why this is private

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+        for key in self.op_impls:
+            impls.remove(key)
+        del fragments_for_existing_libraries[id(self)]
+        del self.m
```
Wondering if we actually need an explicit removal function. Shouldn't just deleting the library be enough?

anjali411 replied:
no because you also want to remove the entry from `fragments_for_existing_libraries`

ezyang commented on torch/custom_libraries.py:
```diff
@@ -0,0 +1,52 @@
+        key = self.ns + "/" + name + "/" + dispatch_key
+        if key in impls:
+            raise RuntimeError("This is not allowed since there's already a kernel overriding " \
+                            + str(getattr(getattr(torch.ops, self.ns), name)) + "'s behavior for " + dispatch_key + " dispatch key.")
```
There's already logic for testing if you are overriding a preexisting kernel in aten/src/ATen/core/dispatch/OperatorEntry.cpp although it's only a warning. Maybe we should just have an option to turn that warning into an error, rather than replicating the logic here?

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -122,6 +190,12 @@ void initDispatchBindings(PyObject* module) {
+    .def("impl", [](py::object self, const char* name, const char* dispatch, py::object func) {
+      self.cast<torch::Library&>().impl(
+        name,
+        dispatch_str(dispatch, CppFunction::makeFromBoxedFunctor(std::make_unique<PythonKernelHolder>(std::move(func))))
```
I think we should set the debug string on the CppFunction here, so that C++ can give a better error message about where a library was defined at. I suggest capturing one frame of the backtrace on Library construction, extracting out the file and the line, and then passing that in as the debug info here.

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

anjali411 commented on torch/library.py:
```diff
@@ -0,0 +1,74 @@
+__all__ = ['extend_library']
+
+# User created fragment libraries to extend existing libraries
+fragments_for_existing_libraries = {}
```
I probably don't need this dictionary

anjali411 replied:
nvm I do need this dictionary to ensure that the library object survives outside the scope of a function

ezyang replied:
Why do you want the library to survive outside of a function? It seems to me that the library should get assigned to a global in a module and live that way.

ezyang commented on torch/_ops.py:
```diff
@@ -61,6 +61,10 @@ def __str__(self):
         return self._overloadpacket
 
+    @property
+    def overloadname(self):
```
I hate this naming, it's so unpythonic, it should be `overload_name` sigh

ezyang replied:
(You shouldn't do this because it's not consistent with what's above)

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,74 @@
+# Every user can create their own fragment to extend existing C++ libraries
+# We don't guarantee the user that another library that they imported is not overriding aten
+# However two libraries are not allowed to override the same operator in the same namespace for the same dispatch key
+def extend_library(ns, dispatch_key=""):
```
extend seems like the wrong word here, since IMPL only lets you add new dispatch key registrations

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,74 @@
+        for key in self.op_impls:
+            impls.remove(key)
+        del fragments_for_existing_libraries[id(self)]
+        del self.m
```
I still think `__del__` should suffice for most of this, I'd rather not have "uninitialized" Library objects floating around (where it's now illegal to do operations on them)

albanD replied:
I think there is one case where we might not want this. If we want to allow to do:
```
returned = torch.ops.aten.sum.default.impl(my_new_sum, "CPU")
```
Basically allowing you to just override the function you want without having to think about library/impl/extend/ etc

So the question is what is returned here? We can say it is a library object and then yes we can bind the lifetime of the function to the library object.
But if we want to allow people to just override it and forget, `returned` can be a handle that can be used to remove it (hook style) otherwise it will live forever.


The argument about having invalid Library object after "remove" is good though.
I guess a good way forward here is:
- Make the lifetime linked to the Library object (so no need for this function)
- We can remove the global dict `_impls_for_existing_libraries` and don't populate it in the init.
- remove the `extend_library` function altogether and simply make the `kind` argument optional for `Library` with a default to `IMPL`.

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,74 @@
+        print("--------", name, self.ns)
+        key = self.ns + "/" + name + "/" + dispatch_key
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
```
so does this work now?

anjali411 replied:
no

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,74 @@
+
+    def impl(self, op_name, dispatch_key, fn):
+        if isinstance(op_name, str):
+            name = self.ns + '.' + op_name
```
If the ns was put on the C++ object it doesn't seem like this concatenation should be necessary

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

ezyang commented on torch/library.h:
```diff
@@ -555,7 +555,7 @@ class TORCH_API Library final {
   Library(Library&&) = default;
   Library& operator=(Library&&) = default;
-
+  // add gen_python_error boolean flag
```
Is this TODO resolved yet?

anjali411 replied:
I'll resolve it in a follow-up PR

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -130,13 +156,15 @@ void initDispatchBindings(PyObject* module) {
       std::move(name),
       std::string(dispatch) == "" ? c10::nullopt : c10::make_optional(c10::parseDispatchKey(dispatch)),
-      "/dev/null",
+      file_and_line,
```
this looks wrong, you have no assurance that the passed in `const char*` has static lifetime; if it gets deallocated you are sadface

anjali411 replied:
would passing the file name as a string and doing `reinterpret_cast<const char*>(std::move(file))` fix the issue?

ezyang replied:
nope

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -11,6 +11,77 @@
+
+        # # Validate that the old behavior is restored for neg and mul
+        # self.assertTrue(not torch.neg(x).is_neg())
+        # self.assertTrue(torch.mul(x, y)._is_zerotensor())
```
clean up?

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -52,6 +51,91 @@ using namespace at;
 using namespace torch::autograd;
 
+std::pair<py::object, py::dict> get_args_kwargs(const c10::OperatorHandle& op, const std::vector<c10::IValue>& arguments) {
+  py::gil_scoped_acquire g;
```
it feels wrong to acquire GIL here, it should be precondition that GIL is held before you call this function

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -52,6 +51,91 @@ using namespace at;
+    const auto& arg = schema.arguments()[idx];
+    kwargs[py::cast(arg.name())] = torch::jit::toPyObject(arguments[idx]);
+  }
+  return std::make_pair(args, kwargs);
```
move into the pair, save a refcount

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -52,6 +51,91 @@ using namespace at;
+    torch::jit::Stack* stack,
+    py::object out,
+    const char* msg) {
+  py::gil_scoped_acquire g;
```
ditto here

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1860,41 +1892,19 @@ void concrete_dispatch_fn(
-  }
+  auto args_kwargs = get_args_kwargs(op, arguments);
+  auto args = args_kwargs.first;
+  auto kwargs = args_kwargs.second;
```
these incur unnecessary refcount bumps, move out of first/second

ezyang commented on torch/csrc/autograd/python_variable.h:
```diff
@@ -62,3 +66,7 @@ inline const at::Tensor& THPVariable_Unpack(PyObject* obj) {
+
+std::pair<py::object, py::dict> get_args_kwargs(const c10::OperatorHandle& op, const std::vector<c10::IValue>& arguments);
+
+void push_out_to_stack(const c10::OperatorHandle& op, torch::jit::Stack* stack, py::object out, const char* msg);
```
The names here feel too generic; it's especially bad because this file is dumping the functions into top-level namespace (no sub-namespace!!) At the very least they should indicate something Python-related is going on. I suggest `parseIValuesToPyArgsKwargs` and `pushPyOutToStack`

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,65 @@
+# The keys in the set are of the form `namespace + "/" + op_name + "/" + dispatch_key`.
+# This set is maintained to ensure that two libraries don't try to override the exact same functionality to avoid
+# libraries calling into kernels not intended to be called.
+impls: Set[str] = set()
```
These should probably get leading underscores as they are not part of the public API for the module

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,65 @@
+# However two libraries are not allowed to override the same operator in the same namespace for the same dispatch key
+def extend_library(ns, dispatch_key=""):
+    '''
+    something
```
TODO

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


TODO: add docs for `torch.extend_library`

[ghstack-poisoned]

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,88 @@
+# User created libraries to extend existing libraries
+# Each user created library is added here to ensure that it's not automatically removed outside the
+# scope of the function it was created in.
+impls_for_existing_libraries = {}
```
this is supposed to be private too right

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,88 @@
+class Library:
+    # kind can be DEF, IMPL
+    def __init__(self, kind, ns, dispatch_key="", message=""):
+        frame = traceback.extract_stack()[0]
```
You're not using the entire stack, only the first frame, so you should set `limit=`

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,88 @@
+    def __init__(self, kind, ns, dispatch_key="", message=""):
+        frame = traceback.extract_stack()[0]
+        filename, lineno = frame.filename, frame.lineno
+        self.m = C._dispatch_library(kind, ns, dispatch_key, filename, lineno) # type: ignore[attr-defined]
```
The correct way to suppress this mypy error is to add `_dispatch_library` to `torch/_C/__init__.pyi.in`

ezyang commented on torch/library.py:
```diff
@@ -0,0 +1,88 @@
+        if key in _impls:
+            raise RuntimeError("This is not allowed since there's already a kernel overriding {}"
+                               "'s behavior for {} dispatch key and {} namespace.".
+                               format(name.split("::")[-1], dispatch_key, self.ns))
```
Technically, you are only checking for a Python conflict; you may still be overriding something in C++. It probably is a good idea to ensure the error message reflects this. Also, it's helpful to print out where the other registration came from (this is what the corresponding C++ code does)

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -51,6 +51,91 @@ using namespace at;
+    const auto& arg = schema.arguments()[idx];
+    kwargs[py::cast(arg.name())] = torch::jit::toPyObject(arguments[idx]);
+  }
+  return std::make_pair(py::cast<py::object>(std::move(args.release())), py::cast<py::dict>(std::move(kwargs.release())));
```
I really feel like `return std::make_pair(std::move(args), std::move(kwargs));` would be enough. I would expect this leaks memory?
btw what is `py::cast<py::object>(py::handle obj)` does wrt refcount?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1751,15 +1834,13 @@ void concrete_dispatch_fn(
   // TODO: put this into the library
   const auto& qualified_name = op.operator_name().name;
-  const auto& overload_name = schema.overload_name();
+  const auto& overload_name = op.schema().overload_name();
```
Why that change? The schema is still in the scope?

albanD commented on torch/library.py:
```diff
@@ -0,0 +1,87 @@
+_impls: Set[str] = set()
+
+class _Library:
+    # kind can be DEF, IMPL
```
Union?
Also I would expect the check below if the kind if valid to happen before we create the c++ object

anjali411 replied:
yes that's what we do

albanD replied:
Isn;t the init function doing:
```python
        # stuff
        self.m = C._dispatch_library(kind, ns, dispatch_key, filename, lineno)
        # more stuff
        if kind != "IMPL":
            raise ValueError("Unsupported kind: ", kind)
```

So that is checked after?

albanD commented on torch/library.py:
```diff
@@ -0,0 +1,87 @@
+            raise ValueError("Unsupported kind: ", kind)
+
+    def __repr__(self):
+        return "<Library(kind='{}', ns='{}', dispatch_key='{}')>".format(self.kind, self.ns, self.dispatch_key)
```
nit: fstring

albanD commented on torch/library.py:
```diff
@@ -0,0 +1,87 @@
+    def __repr__(self):
+        return "<Library(kind='{}', ns='{}', dispatch_key='{}')>".format(self.kind, self.ns, self.dispatch_key)
+
+    def impl(self, op_name, fn, dispatch_key=''):
```
doc?

albanD commented on torch/library.py:
```diff
@@ -0,0 +1,87 @@
+# libraries calling into kernels not intended to be called.
+_impls: Set[str] = set()
+
+class _Library:
```
Not sure why this class is private tbh. We return it and expect the user to manipulate it.

albanD commented on torch/library.py:
```diff
@@ -0,0 +1,87 @@
+        else:
+            raise RuntimeError("impl should be passed either a name or an OpOverload object as the first argument")
+
+        key = self.ns + "/" + name.split("::")[-1] + "/" + dispatch_key
```
nit: key could just be a tuple here to avoid string manipulation no?

albanD commented on test/test_python_dispatch.py:
```diff
@@ -12,6 +12,86 @@
+
+        self.assertTrue(torch.neg(x).is_neg())
+
+        # RuntimeError: impl("aten::neg", ...):
```
Not sure what this comment is for?

anjali411 replied:
just the c++ error

albanD commented on test/test_python_dispatch.py:
```diff
@@ -12,6 +12,86 @@
+            return args[0]
+
+        my_lib1 = torch.library.extend_library("aten")
+        my_lib1.impl('aten::sum', my_sum, "CPU")
```
We really need to specify what kind of string is allowed for `op_name` here. It seems that both `sum` and `aten::sum` works based on the tests? But if I pass `foo::sum`, will that work?

anjali411 replied:
If foo is a valid namespace, then it should work

albanD replied:
Even though the library was created for the namespace "aten"?

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

anjali411 committed:
Update on "Allow users to override kernels for existing C++ ops through Python"


API:
```
def my_sum(*args, **kwargs):
    return args[0]
my_lib1 = torch.library.extend_library("aten")
my_lib1.impl('aten::sum', my_sum)
x = torch.tensor([1, 2])
my_lib1.remove()
```

[ghstack-poisoned]

anjali411 commented:
@pytorchbot merge this please

malfet commented:
This PR breaks internal build with the following error message:
```
xplat/caffe2/torch/csrc/utils/python_dispatch.cpp:58:3: error: unknown type name 'SafePyObject'; did you mean 'at::SafePyObject'?
  SafePyObject func_;
```

but alas, I can not revert it, need to submit a quick forward fix
